### PR TITLE
add threesdk formula v1.0

### PR DESCRIPTION
**Pull request by waleedhammam**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/threesdk.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+class Threesdk < Formula
+  desc "Software development kit for JumpscaleX"
+  homepage "https://wiki.threefold.io/"
+  url "https://github.com/threefoldtech/jumpscaleX_core/archive/v10.5.2.tar.gz"
+  sha256 "d2ce77a61a198b2cd53f4acce5fd7e5c36c369fc83a05b9f74314530b4f4a175"
+
+  bottle do
```
Comment (SMillerDev) [inline Formula/threesdk.rb:7]: New formula should not have a bottle block.

File: Formula/threesdk.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    sha256 "c71db15326ee9196cd98602e38d0b7fb2b818cdd48eede4ee8eb827d809e09ba" => :el_capitan
+    sha256 "85cc828a96735bdafcf29eb6291ca91bac846579bcef7308536e0c875d6c81d7" => :yosemite
+  end
+
+  depends_on "patchelf" => :install
+  depends_on "python" => :install
+  depends_on "upx" => :install
+
+  def install
+    ENV["PATH"] += ":/usr/local/bin:/home/linuxbrew/.linuxbrew/bin"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:22]: This is not needed

File: Formula/threesdk.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    system "cd install; pip3 install -r requirements.txt"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:23]: Resources should be downloaded with a pinned version and hash. Please look at what other python installs do.

File: Formula/threesdk.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    sha256 "ef65c759c5097a36323fa9c77756468649e8d1980a3a4e05695c05e39568967c" => :catalina
+    sha256 "28f4090610946a4eb207df102d841de23ced0d06ba31cb79e040d883906dcd4f" => :mojave
+    sha256 "91dd0caca9bd3f38c439d5a7b6f68440c4274945615fae035ff0a369264b8a2f" => :high_sierra
+    sha256 "4921af80137af9cc3d38fd17c9120da882448a090b0a8a3a19af3199b415bfca" => :sierra
+    sha256 "c71db15326ee9196cd98602e38d0b7fb2b818cdd48eede4ee8eb827d809e09ba" => :el_capitan
+    sha256 "85cc828a96735bdafcf29eb6291ca91bac846579bcef7308536e0c875d6c81d7" => :yosemite
+  end
+
+  depends_on "patchelf" => :install
+  depends_on "python" => :install
```
Comment (SMillerDev) [inline Formula/threesdk.rb:18]: Please use `python@3.8`

File: Formula/threesdk.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  depends_on "upx" => :install
+
+  def install
+    ENV["PATH"] += ":/usr/local/bin:/home/linuxbrew/.linuxbrew/bin"
+    system "cd install; pip3 install -r requirements.txt"
+    system "cd install; ./package.sh"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:24]: Does this download anything?

File: Formula/threesdk.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    system "cd install; cp dist/3sdk /usr/local/bin/3sdk;"
+  end
+
+  test do
+    system "false"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:29]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

**Comment by SMillerDev** [comment]
There is no way this passed all the commands that you checked the box for in the PR template.

File: Formula/threesdk.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    sha256 "28f4090610946a4eb207df102d841de23ced0d06ba31cb79e040d883906dcd4f" => :mojave
+    sha256 "91dd0caca9bd3f38c439d5a7b6f68440c4274945615fae035ff0a369264b8a2f" => :high_sierra
+    sha256 "4921af80137af9cc3d38fd17c9120da882448a090b0a8a3a19af3199b415bfca" => :sierra
+    sha256 "c71db15326ee9196cd98602e38d0b7fb2b818cdd48eede4ee8eb827d809e09ba" => :el_capitan
+    sha256 "85cc828a96735bdafcf29eb6291ca91bac846579bcef7308536e0c875d6c81d7" => :yosemite
+  end
+
+  depends_on "patchelf" => :install
+  depends_on "python" => :install
+  depends_on "upx" => :install
```
Comment (bayandin) [inline Formula/threesdk.rb:19]: We don't have `:install` in any core formula, where did it come from?

```suggestion
  depends_on "patchelf" => :build
  depends_on "python@3.8" => :build
  depends_on "upx" => :build
```

File: Formula/threesdk.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+class Threesdk < Formula
+  desc "Software development kit for JumpscaleX"
+  homepage "https://wiki.threefold.io/"
```
Comment (bayandin) [inline Formula/threesdk.rb:3]: This homepage doesn't look related to "Threesdk"

File: Formula/threesdk.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    rebuild 1
+    sha256 "ef65c759c5097a36323fa9c77756468649e8d1980a3a4e05695c05e39568967c" => :catalina
+    sha256 "28f4090610946a4eb207df102d841de23ced0d06ba31cb79e040d883906dcd4f" => :mojave
+    sha256 "91dd0caca9bd3f38c439d5a7b6f68440c4274945615fae035ff0a369264b8a2f" => :high_sierra
+    sha256 "4921af80137af9cc3d38fd17c9120da882448a090b0a8a3a19af3199b415bfca" => :sierra
+    sha256 "c71db15326ee9196cd98602e38d0b7fb2b818cdd48eede4ee8eb827d809e09ba" => :el_capitan
+    sha256 "85cc828a96735bdafcf29eb6291ca91bac846579bcef7308536e0c875d6c81d7" => :yosemite
+  end
+
+  depends_on "patchelf" => :install
```
Comment (SMillerDev) [inline Formula/threesdk.rb:17]: This is only needed on linux

File: Formula/threesdk.rb (hunk 5 above)
Comment (SMillerDev) [reply Formula/threesdk.rb:24]: it runs brew commands, this will need to be replaced by the 2 commands it actually runs:
```ruby
inreplace "threesdk/__init__.py" "_unreleased_" version.to_s
system "pyinstaller 3sdk.spec"
```

File: Formula/threesdk.rb (hunk 1 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: Removed now

File: Formula/threesdk.rb (hunk 2 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: removed now

File: Formula/threesdk.rb (hunk 3 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: Add all the resources with hash and version

File: Formula/threesdk.rb (hunk 4 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: replaced it, also :build instead of :install

File: Formula/threesdk.rb (hunk 5 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: done, thanks for the note

File: Formula/threesdk.rb (hunk 6 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: add a test for checking the running containers, if docker is not installed it will raise error, if installed it will show empty result, if a 3sdk container is deployed it'll appear in the list

File: Formula/threesdk.rb (hunk 7 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: replaced, thanks

File: Formula/threesdk.rb (hunk 8 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: this was the wiki for it but replaced with the direct one

File: Formula/threesdk.rb (hunk 9 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: removed

File: Formula/threesdk.rb [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    end
+
+    resource "PyNaCl" do
+      url "https://files.pythonhosted.org/packages/61/ab/2ac6dea8489fa713e2b4c6c5b549cc962dd4a842b5998d9e80cf8440b7cd/PyNaCl-1.3.0.tar.gz"
+      sha256 "0c6100edd16fefd1557da078c7a31e7b7d7a52ce39fdca2bec29d4f7b6e7600c"
+    end
+
+    def install
+      system 'cd install; version=$(git describe --tags); gsed -i "s/_unreleased_/${version}/" threesdk/__init__.py;'
+      system "cd install; pyinstaller 3sdk.spec"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:133]:
```suggestion
      chdir "install" do
        inreplace "threesdk/__init__.py", "_unreleased_" version.to_s
        system "pyinstaller 3sdk.spec"
      end
```

File: Formula/threesdk.rb [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      bin.install "install/dist/3sdk"
+    end
+
+    test do
+      system bin/"3sdk", "--expert", "container", "list"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:138]: Can you check for some output here?

File: Formula/threesdk.rb [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    desc "Software development kit for JumpscaleX"
+    homepage "https://sdk.threefold.io/"
+    url "https://github.com/threefoldtech/jumpscaleX_core/archive/v10.5.2.tar.gz"
+    sha256 "d2ce77a61a198b2cd53f4acce5fd7e5c36c369fc83a05b9f74314530b4f4a175"
+
+    depends_on "python@3.8" => :build
+    depends_on "gnu-sed" => :build
+    depends_on "upx" => :build
+
+    resource "docopt" do
```
Comment (SMillerDev) [inline Formula/threesdk.rb:11]: It doesn't install any of these

File: Formula/threesdk.rb [hunk 13]
```diff
@@ -0,0 +1,140 @@
+class Threesdk < Formula
+    desc "Software development kit for JumpscaleX"
+    homepage "https://sdk.threefold.io/"
+    url "https://github.com/threefoldtech/jumpscaleX_core/archive/v10.5.2.tar.gz"
+    sha256 "d2ce77a61a198b2cd53f4acce5fd7e5c36c369fc83a05b9f74314530b4f4a175"
+
+    depends_on "python@3.8" => :build
```
Comment (SMillerDev) [inline Formula/threesdk.rb:7]: Is this really only needed to build the software?

File: Formula/threesdk.rb (hunk 10 above)
Comment (waleedhammam) [reply Formula/threesdk.rb:133]: `inreplace "threesdk/__init__.py", "_unreleased_", version.to_s`
this fails with me with this error
image

File: Formula/threesdk.rb (hunk 13 above)
Comment (waleedhammam) [reply Formula/threesdk.rb:7]: yes using pyinstaller to package the binary

File: Formula/threesdk.rb (hunk 12 above)
Comment (waleedhammam) [reply Formula/threesdk.rb:11]: sorry didn't get that, we need upx, but gnu-sed can be removed, That what you mean ?

File: Formula/threesdk.rb (hunk 11 above)
Comment (waleedhammam) [reply Formula/threesdk.rb:138]: will do and check for output

File: Formula/threesdk.rb (hunk 12 above)
Comment (SMillerDev) [reply Formula/threesdk.rb:11]: None of the resources are installed

File: Formula/threesdk.rb (hunk 10 above)
Comment (SMillerDev) [reply Formula/threesdk.rb:133]: There's probably more info around this error.

File: Formula/threesdk.rb (hunk 13 above)
Comment (SMillerDev) [reply Formula/threesdk.rb:7]: in that case it can just use 
```suggestion
    depends_on "pyinstaller" => :build
```

Comment (waleedhammam) [reply Formula/threesdk.rb:7]: Then no need for it at all pyinstaller is installed from the resources, it's a pip

File: Formula/threesdk.rb (hunk 10 above)
Comment (waleedhammam) [reply Formula/threesdk.rb:133]: sadly no, this is the only related part
image

File: Formula/threesdk.rb (hunk 13 above)
Comment (SMillerDev) [reply Formula/threesdk.rb:7]: If something is installable through homebrew you must use that version, rather than resourcing your own.

File: Formula/threesdk.rb (hunk 10 above)
Comment (SMillerDev) [reply Formula/threesdk.rb:133]: Apparently you don't need the inreplace since their released tarbal doesn't have that part of the script.

File: Formula/threesdk.rb [hunk 14]
```diff
@@ -0,0 +1,140 @@
+class Threesdk < Formula
```
Comment (SMillerDev) [inline Formula/threesdk.rb:1]: Since the repo is called `jumpScaleX_core` Shouldn't the formula be called `JumpScaleCore`? Why the different name?

File: Formula/threesdk.rb (hunk 13 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: yes add it with the hash and md5

File: Formula/threesdk.rb (hunk 12 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: installed now using virual env

File: Formula/threesdk.rb (hunk 10 above)
Comment (waleedhammam) [reply Formula/threesdk.rb]: gnused is still working

File: Formula/threesdk.rb (hunk 14 above)
Comment (waleedhammam) [reply Formula/threesdk.rb:1]: just that it's the commercial name of our software

Comment (SMillerDev) [reply Formula/threesdk.rb:1]: Either rename the repo or rename the formula.

File: Formula/threesdk.rb [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+
+  resource "pycparser" do
+    url "https://files.pythonhosted.org/packages/0f/86/e19659527668d70be91d0369aeaa055b4eb396b0f387a4f92293a20035bd/pycparser-2.20.tar.gz"
+    sha256 "2d475327684562c3a96cc71adf7dc8c4f0565175cf86b6d7a404ff4c771f15f0"
+  end
+
+  resource "PyNaCl" do
+    url "https://files.pythonhosted.org/packages/61/ab/2ac6dea8489fa713e2b4c6c5b549cc962dd4a842b5998d9e80cf8440b7cd/PyNaCl-1.3.0.tar.gz"
+    sha256 "0c6100edd16fefd1557da078c7a31e7b7d7a52ce39fdca2bec29d4f7b6e7600c"
+  end
```
Comment (SMillerDev) [inline Formula/threesdk.rb:130]: This is not acceptable, pyinstaller is available as a formula and should be used as `depends_on`

File: Formula/threesdk.rb [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    venv = virtualenv_create(libexec, "python3")
+    %w[
+      docopt jedi parso prompt-toolkit ptpython Pygments six wcwidth certifi
+      chardet idna requests urllib3 altgraph macholib PyInstaller packaging
+      pyparsing pudb urwid cffi pycparser PyNaCl
+    ].each do |r|
+      venv.pip_install resource(r)
+    end
+    chdir "install" do
+      ENV["PATH"] += ":/usr/local/bin" # needed for python
```
Comment (SMillerDev) [inline Formula/threesdk.rb:143]: This is not needed for python, hunderds of formula run fine without it.

File: Formula/threesdk.rb [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+      # inreplace doesn't work
+      system 'version=$(git describe --tags); gsed -i "s/_unreleased_/${version}/" threesdk/__init__.py;'
```
Comment (SMillerDev) [inline Formula/threesdk.rb:145]: You're not doing a git checkout so this doesn't work. This whole line isn't needed since `_unreleased_` in the file in the archive

File: Formula/threesdk.rb [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+      bin.install "dist/3sdk"
+    end
+  end
+
+  test do
+    # make sure binary is running ok
+    system bin / "3sdk", "--expert", "info"
+    assert_match "builder: manage container images", shell_output("#{bin}/3sdk info").strip
+    # install requires docker to be installed by user if not it will fail
+    system bin / "3sdk container install explorer=none secret=123456"
```
Comment (SMillerDev) [inline Formula/threesdk.rb:156]:
```suggestion
    system bin/"3sdk container install explorer=none secret=123456"
```

File: Formula/threesdk.rb [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+      system "pyinstaller", "3sdk.spec"
+      bin.install "dist/3sdk"
+    end
+  end
+
+  test do
+    # make sure binary is running ok
+    system bin / "3sdk", "--expert", "info"
+    assert_match "builder: manage container images", shell_output("#{bin}/3sdk info").strip
+    # install requires docker to be installed by user if not it will fail
```
Comment (SMillerDev) [inline Formula/threesdk.rb:155]: We don't have docker in CI so this should expect to fail.

**Comment by SMillerDev** [comment]
Please follow the directions about pyinstaller this time, you're already violating this rule: https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff and if I have to spend too much time repeating myself I'll just decline the request instead.
